shmem.c

```diff
@@ -194,22 +267,56 @@ static void* ashv_thread_function(void* arg)
 	while ((sendsock = accept(sock, (struct sockaddr *)&addr, &len)) != -1) {
 		int shmid;
 		if (recv(sendsock, &shmid, sizeof(shmid), 0) != sizeof(shmid)) {
-			DBG("%s: ERROR: recv() returned not %zu bytes", __PRETTY_FUNCTION__, sizeof(shmid));
+			DBG("%s: ERROR: recv() returned not %zu shmid bytes", __PRETTY_FUNCTION__, sizeof(shmid));
+			close(sendsock);
+			continue;
+		}
+		int action;
+		if (recv(sendsock, &action, sizeof(action), 0) != sizeof(action)) {
+			DBG("%s: ERROR: recv() returned not %zu action bytes", __PRETTY_FUNCTION__, sizeof(action));
 			close(sendsock);
 			continue;
 		}
 		pthread_mutex_lock(&mutex);
 		int idx = ashv_find_local_index(shmid);
 		if (idx != -1) {
-			if (write(sendsock, &shmem[idx].key, sizeof(key_t)) != sizeof(key_t)) {
-				DBG("%s: ERROR: write failed: %s", __PRETTY_FUNCTION__, strerror(errno));
-			}
-			if (ancil_send_fd(sendsock, shmem[idx].descriptor) != 0) {
-				DBG("%s: ERROR: ancil_send_fd() failed: %s", __PRETTY_FUNCTION__, strerror(errno));
+			pid_t pid;
+			switch (action) {
+			case ASHV_RM:
+				DBG("%s: action ASHV_RM(%d) for %d", __PRETTY_FUNCTION__, action, shmid);
+				ashv_delete_segment(idx);
+				break;
+			case ASHV_GET:
+				DBG("%s: action ASHV_GET(%d) for %d", __PRETTY_FUNCTION__, action, shmid);
+				SOCKET_ASHV_WRITE(key_t, key)
+				SOCKET_ASHV_WRITE(bool, markedForDeletion)
+				SOCKET_ASHV_WRITE(int, countAttach)
+				if (ancil_send_fd(sendsock, shmem[idx].descriptor) != 0)
+					DBG("%s: ERROR: ancil_send_fd() failed: %s", __PRETTY_FUNCTION__, strerror(errno));
+				break;
+			case ASHV_AT:
+				DBG("%s: action ASHV_AT(%d) for %d", __PRETTY_FUNCTION__, action, shmid);
+				SOCKET_ASHV_READ(pid_t, pid)
+				android_shmem_attach_pid(idx, pid);
+				break;
+			case ASHV_DT:
+				DBG("%s: action ASHV_DT(%d) for %d", __PRETTY_FUNCTION__, action, shmid);
+				SOCKET_ASHV_READ(pid_t, pid)
+				android_shmem_detach_pid(idx, pid);
+				break;
+			case ASHV_UPD:
+				DBG("%s: action ASHV_UPD(%d) for %d", __PRETTY_FUNCTION__, action, shmid);
+				SOCKET_ASHV_WRITE(bool, markedForDeletion)
+				android_shmem_check_pids(idx);
+				SOCKET_ASHV_WRITE(int, countAttach)
+				ashv_write_pids(sendsock, idx);
+				break;
+			default:
+				DBG("%s: ERROR: unknown action %d", __PRETTY_FUNCTION__, action);
+				break;
 			}
-		} else {
```

Comment:
There are several places in this PR where you removed the braces from a single-line block unrelated to this change. Please put them back to keep this change minimal, and please make one pass through your new code to make sure all new single-line blocks have braces to match the style of this file (among other reasons).